src/bytes.rs

```diff
@@ -71,14 +101,73 @@ impl<'a> Bytes<'a> {
             len: data.len(),
             data: data.as_ptr().cast(),
             capacity: data.len(),
-            vtable: &VT,
+            vtable: unsafe {
+                NonNull::new_unchecked(&VT as &'static BytesVt as *const BytesVt as *mut BytesVt)
+            },
+            marker: core::marker::PhantomData,
         }
     }
+
+    /// Constructs a [`Bytes`] referring to a slice.
+    ///
+    /// Unlike [`Bytes::from_static`] and `<Bytes as From<&'static ..>>::from` implementations, the resulting [`Bytes`] cannot tell that it's referring to `'static` data even if it is, and will therefore perform a copy when [`Bytes::upgrade`]ing.
+    /// ```
+    /// # use safer_ffi::bytes::Bytes;
+    /// let data = "Hello there, this string is long enough that it'll cross the inline-threshold (core::mem::size_of::<Bytes>() - 1) on all supported platforms";
+    /// let mut bytes = Bytes::from_slice(data.as_bytes());
+    /// assert!(bytes.upgrade_will_allocate());
+    /// let data = "This string isn't";
+    /// let mut bytes = Bytes::from_slice(data.as_bytes());
+    /// assert!(!bytes.upgrade_will_allocate());
+    /// ```
+    ///
+    /// Note that if the slice is small enough to be interned in the [`Bytes`], it'll be, allowing for free upgrades. You may use the [`Bytes::intern_slice`] constructor instead of this one if you want to be able to handle interning not being possible.
+    pub fn from_slice(data: &'a [u8]) -> Self {
+        static VT: BytesVt = BytesVt {
+            release: None,
+            retain: Some(noop),
+        };
+        Self::intern_slice(data).unwrap_or_else(|| Self {
+            start: data.as_ptr().cast(),
+            len: data.len(),
+            data: data.as_ptr().cast(),
+            capacity: data.len(),
+            vtable: unsafe {
+                NonNull::new_unchecked(&VT as &'static BytesVt as *const BytesVt as *mut BytesVt)
+            },
+            marker: core::marker::PhantomData,
+        })
+    }
+
+    /// Constructs a [`Bytes`] from a short slice by interning it, untying [`Bytes`]'s lifetime from `slice`'s.
+    ///
+    /// If the slice is too long to be interned, this constructor returns `None` instead.
+    ///
+    /// A slice may be interned if it is at least one byte smaller than [`Bytes`], which has the same size as `[usize; 5]`.
+    ///
+    /// ```
+    /// # use safer_ffi::bytes::Bytes;
+    /// let interned = Bytes::intern_slice("Hi".as_bytes()).unwrap();
+    /// assert_eq!(interned.as_slice(), "Hi".as_bytes());
+    /// assert!(!interned.clone_will_allocate());
+    /// assert!(!interned.upgrade_will_allocate());
+    /// assert!(Bytes::intern_slice("This slice is too long to intern, even on architectures with 128 bit pointer-size.".as_bytes()).is_none())
+    /// ```
+    pub fn intern_slice(slice: &[u8]) -> Option<Bytes<'static>> {
+        let mut buffer = [0; core::mem::size_of::<Self>()];
+        if slice.len() >= buffer.len() {
```

Comment:
# ⚠️ About endianness ⚠️

Actually, I don't think you can use `mem::size_of::<Self>()`, here, **unconditionally**.

Indeed, on a big-endian architecture, especially one using 32-bit-addressing, the "alignment" of the vtable address may end up as the _first_ byte of `.vtable`, rather than the last.

```rs
const IS_LITTLE_ENDIAN: bool =
    0x42_u16.to_ne_bytes() == 0x42_u16.to_le_bytes()
;
const MAX_INLINE_SIZE: usize =
    mem::size_of::<Self>()
        - if IS_LITTLE_ENDIAN { mem::size_of::<usize>() } else { 0 }
        - 1 // the length and discriminant packing byte.
;
```

so as to, here:

```suggestion
        let mut buffer = [0; core::mem::size_of::<Self>()];
        if slice.len() > MAX_INLINE_SIZE {
```

Reply:
damn, you're right! D: